src/tempo.py

```diff
@@ -24,57 +24,56 @@ class Tempo:
 
-    # todo make configurable?
+    server_ports = {
+        "tempo_http": 3200,
+        # "tempo_grpc": 9096, # default grpc listen port is 9095, but that conflicts with promtail.
+    }
+
     receiver_ports: Dict[ReceiverProtocol, int] = {
         "zipkin": 9411,
-        "tempo_http": -1,  # configurable; populated by __init__
-        "tempo_grpc": -1,  # configurable; populated by __init__
-        "tempo": 3201,  # legacy, renamed to tempo_http
         "otlp_grpc": 4317,
         "otlp_http": 4318,
-        "jaeger_grpc": 14250,
         "jaeger_thrift_http": 14268,
-        "jaeger_http_thrift": 14269,  # legacy, renamed to jaeger_thrift_http
-        # todo add support for:
+        # todo if necessary add support for:
         #  "kafka": 42,
+        #  "jaeger_grpc": 14250,
         #  "opencensus": 43,
         #  "jaeger_thrift_compact": 44,
         #  "jaeger_thrift_binary": 45,
     }
 
+    all_ports = {**server_ports, **receiver_ports}
+
     def __init__(
         self,
         container: ops.Container,
-        http_port: int = 3200,
-        grpc_port: int = 9096,
         local_host: str = "0.0.0.0",
         enable_receivers: Optional[Sequence[ReceiverProtocol]] = None,
     ):
-        self.receiver_ports["tempo_http"] = http_port
-        # default grpc listen port is 9095, but that conflicts with promtail.
-        self.receiver_ports["tempo_grpc"] = grpc_port
-
         # ports source: https://github.com/grafana/tempo/blob/main/example/docker-compose/local/docker-compose.yaml
         self._local_hostname = local_host
         self.container = container
-        self._enabled_receivers = enable_receivers or []
-        self._supported_receivers: Tuple[ReceiverProtocol, ...] = tuple(self.receiver_ports)
+        self.enabled_receivers = enable_receivers or []
 
     @property
-    def tempo_port(self) -> int:
+    def tempo_server_port(self) -> int:
         """Return the receiver port for the built-in tempo_http protocol."""
-        return self.receiver_ports["tempo_http"]
+        return self.server_ports["tempo_http"]
+
+    def get_external_ports(self, service_name_prefix: str) -> List[Tuple[str, int, int]]:
+        """List of service names and port mappings for the kubernetes service patch.
 
-    def get_ports(self, service_name_prefix: str) -> List[Tuple[str, int, int]]:
-        """List of service names and port mappings for the kubernetes service patch."""
+        Includes the tempo server as well as the receiver ports.
+        """
         # todo allow remapping ports?
+        all_ports = {**self.server_ports}
         return [
             (
-                (f"{service_name_prefix}-{protocol}").replace("_", "-"),
-                self.receiver_ports[protocol],
-                self.receiver_ports[protocol],
+                (f"{service_name_prefix}-{service_name}").replace("_", "-"),
+                all_ports[service_name],
+                all_ports[service_name],
```

Comment:
Might be a stupid question but now that I think about it: why do we have the same int argument twice here in the tuple?

Reply:
IIRC the kubernetesservicepatch thingy (which this is for) takes two ports to allow remapping them.